### ISSUE TITLE
Infinite recursion when hovering over class definition with duplicate names

### ISSUE THREAD

**Issue by Jaakkonen** [microsoft/pyright#2314]
**Describe the bug**
```sh
# Foo.py
class Foo: ...
# bar.py
from . import Foo
class Foo(Foo.Foo): ...
```
Hovering over the new class in `bar.py` causes a recursion that prints 
```
[Error - 5:58:08 PM] Request textDocument/hover failed.
  Message: Request textDocument/hover failed with message: Maximum call stack size exceeded
  Code: -32603 
```
to pyright vscode extension output.

**To Reproduce**
Create files with content described above, open `bar.py` on VS Code with pyright extension and hower over `class Foo` part.

**Expected behavior**
No recursion errors :)

**VS Code extension or command-line**
Extension version v1.1.169 from latest git main. VS Code from latest git main.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
The symbol `Foo` in the scope of the `bar` module is assigned twice — once in the `import` statement and a second time in the `class` statement. The go-to-definition behavior is therefore expected in this case. It's showing you all of the places where that symbol are defined.

I agree that the hover issue is a bug, but it's not related to the first bug. It's due to some special-case handling we have in the hover provider to handle `from .A import A` within an `__init__.py` file. This special-case logic shouldn't apply in this case and is giving the incorrect hover result.

In general, I would encourage you to avoid name conflicts like the one in this example. It will improve the readability, maintainability, and robustness of your code. You can use an `as` clause in the import statement. Or follow the standard capitalization conventions for python modules and classes.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.170, which I just published. It will also be included in the next release of pylance.
